community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/PageList.java

```diff
@@ -0,0 +1,332 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.io.pagecache.impl.muninn;
+
+import java.io.IOException;
+
+import org.neo4j.io.pagecache.Page;
+import org.neo4j.io.pagecache.PageCursor;
+import org.neo4j.io.pagecache.PageSwapper;
+import org.neo4j.io.pagecache.tracing.PageFaultEvent;
+import org.neo4j.unsafe.impl.internal.dragons.MemoryManager;
+import org.neo4j.unsafe.impl.internal.dragons.UnsafeUtil;
+
+import static java.lang.String.format;
+
+/**
+ * The PageList maintains the off-heap meta-data for the individual memory pages.
+ *
+ * The meta-data for each page is the following:
+ *
+ * <table>
+ *     <tr><th>Bytes</th><th>Use</th></tr>
+ *     <tr><td>8</td><td>Sequence lock word.</td></tr>
+ *     <tr><td>8</td><td>Pointer to the memory page.</td></tr>
+ *     <tr><td>8</td><td>File page id.</td></tr>
+ *     <tr><td>4</td><td>Page swapper id.</td></tr>
+ *     <tr><td>1</td><td>Usage stamp. Optimistically incremented; truncated to a max of 4.</td></tr>
+ *     <tr><td>3</td><td>Padding.</td></tr>
+ * </table>
+ */
+class PageList
+{
+    private static final int META_DATA_BYTES_PER_PAGE = 32;
+    private static final int OFFSET_LOCK_WORD = 0; // 8 bytes
+    private static final int OFFSET_ADDRESS = 8; // 8 bytes
+    private static final int OFFSET_FILE_PAGE_ID = 16; // 8 bytes
+    private static final int OFFSET_SWAPPER_ID = 24; // 4 bytes
+    private static final int OFFSET_USAGE_COUNTER = 28; // 1 byte
+
+    private final long pageCount;
+    private final int cachePageSize;
+    private final MemoryManager memoryManager;
+    private final long victimPageAddress;
+    private final long baseAddress;
+
+    PageList( long pageCount, int cachePageSize, MemoryManager memoryManager, long victimPageAddress )
+    {
+        this.pageCount = pageCount;
+        this.cachePageSize = cachePageSize;
+        this.memoryManager = memoryManager;
+        this.victimPageAddress = victimPageAddress;
+        long bytes = pageCount * META_DATA_BYTES_PER_PAGE;
+        this.baseAddress = memoryManager.allocateAligned( bytes );
+        clearMemory( baseAddress, pageCount );
+    }
+
+    private void clearMemory( long baseAddress, long pageCount )
+    {
+        long address = baseAddress - 8;
+        for ( long i = 0; i < pageCount; i++ )
+        {
+            UnsafeUtil.putLong( address += 8, 0 ); // lock word
+            UnsafeUtil.putLong( address += 8, 0 ); // pointer
+            UnsafeUtil.putLong( address += 8, PageCursor.UNBOUND_PAGE_ID ); // file page id
+            UnsafeUtil.putLong( address += 8, 0 ); // rest
+        }
+        UnsafeUtil.fullFence(); // Guarantee the visibility of the cleared memory
```

Comment:
Wearing belt and suspenders here because I'm using plain writes, and this is happening before the `OffHeapPageLock` takes over the responsibility of protecting the pages and their meta-data.